### PR TITLE
More support for Array and Slice

### LINKED ISSUE #184: Support slice and array indexing
Implement support for [ProjectionElem::Index](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/syntax/enum.ProjectionElem.html#variant.Index). This requires a bit of refactoring to the code in https://github.com/liquid-rust/flux/blob/0520bc222f31fa9c3293f93eebe9088c5031b65f/flux-typeck/src/type_env/paths_tree.rs.

### PR DESCRIPTION
**Pull request by ranjitjhala**
Fixes #184 maybe?

The main big change is factoring `Array` and `Slice` as `BaseTy` to allow refining with size.

### CODE REVIEW

File: flux-desugar/src/desugar.rs [hunk 1]
```diff
@@ -1058,6 +1078,19 @@ impl Binder {
...
 
+    fn from_base(
+        name_gen: &IndexGen<fhir::Name>,
+        map: &fhir::Map,
+        base: &surface::BaseTy<Res>,
+    ) -> Binder {
+        match base {
+            surface::BaseTy::Path(path) => Binder::from_res(name_gen, map, path.ident),
+            surface::BaseTy::Array(_, _) | surface::BaseTy::Slice(_) => {
+                Binder::from_res(name_gen, map, Res::Uint(surface::UintTy::Usize))
```
Comment (nilehmann) [inline flux-desugar/src/desugar.rs:1089]: I would inline this call as
```suggestion
                Binder::Single(name_gen.fresh(), fhir::Sort::Int)
```

File: flux-middle/src/rty/mod.rs [hunk 2]
```diff
@@ -575,6 +567,16 @@ impl BaseTy {
         BaseTy::Adt(adt_def, substs.into())
     }
 
+    pub fn array(ty: Ty, c: Const) -> BaseTy {
+        BaseTy::Array(ty, c)
+        // Ty::indexed(BaseTy::Array(ty, c), RefineArgs::empty())
```
Comment (nilehmann) [inline flux-middle/src/rty/mod.rs:572]:
```suggestion
```

File: flux-middle/src/rty/mod.rs [hunk 3]
```diff
@@ -575,6 +567,16 @@ impl BaseTy {
...
+    }
+
+    pub fn slice(ty: Ty) -> BaseTy {
+        BaseTy::Slice(ty)
+        //Ty::indexed(BaseTy::Slice(ty), RefineArgs::empty())
```
Comment (nilehmann) [inline flux-middle/src/rty/mod.rs:577]:
```suggestion
```

File: flux-desugar/src/desugar.rs [hunk 4]
```diff
@@ -870,6 +878,13 @@ impl Binders {
         Ok(())
     }
 
+    // fn base_res(base: &surface::BaseTy<Res>) -> Res {
+    //     match base {
+    //         surface::BaseTy::Path(path) => path.ident,
+    //         _ => panic!("FIXME"),
+    //     }
+    // }
+
```
Comment (nilehmann) [inline flux-desugar/src/desugar.rs:887]:
```suggestion
```

File: flux-typeck/src/checker.rs [hunk 5]
```diff
@@ -669,14 +682,36 @@ impl<'a, 'tcx, P: Phase> Checker<'a, 'tcx, P> {
...
+        env: &mut TypeEnv,
+        src_info: SourceInfo,
+        place: &Place,
+    ) -> Result<Ty, CheckerError> {
+        let gen = &mut self.phase.constr_gen(self.genv, rcx, Tag::Other);
+        let ty = env
+            .lookup_place(rcx, gen, place)
+            .map_err(|err| CheckerError::from(err).with_src_info(src_info))?;
+        match ty.kind() {
+            TyKind::Indexed(BaseTy::Array(_, _), ixs) | TyKind::Indexed(BaseTy::Slice(_), ixs) => {
```
Comment (nilehmann) [inline flux-typeck/src/checker.rs:705]: We typically use `idxs`

File: flux-typeck/src/checker.rs [hunk 6]
```diff
@@ -669,14 +682,36 @@ impl<'a, 'tcx, P: Phase> Checker<'a, 'tcx, P> {
...
+                Ok(Ty::indexed(BaseTy::Uint(UintTy::Usize), ixs.clone()))
+            }
+            TyKind::Exists(BaseTy::Array(_, _), ixs) | TyKind::Exists(BaseTy::Slice(_), ixs) => {
+                Ok(Ty::exists(BaseTy::Uint(UintTy::Usize), ixs.clone()))
+            }
```
Comment (nilehmann) [inline flux-typeck/src/checker.rs:710]: This case is not necessary. Existentials should always be unpacked before going into the `TypeEnv`, if they are not, then there’s a bug somewhere else.

```suggestion
```

**Review by nilehmann — APPROVED** [review]
My initial reaction was that indexing arrays by the length would cause inconsistencies as you could write an array where the indexed length is different than the const length

```rust
#[flux::sig(fn([i32; _][2]))]
fn myarr(a: [i32; 1]) {
    a[1]
}
```
But if array literal are the only way to construct arrays then I think this should be ok and it just declaring an inhabited type. However, not connecting both lengths makes code cumbersome. For example, it makes it impossible to use the result of a function like https://doc.rust-lang.org/std/primitive.u32.html#method.to_be_bytes without explicitly checking the length, even though you already know the length is 4. Ideally, arrays shouldn’t be indexed but rather we the const length should be automatically _reflected_ into the logic when needed, i.e., the typing rule for `Len` should be something like

```
|- a: [T; n]         reflect(n) = a
———————————————————————————--———-——-
          |- Len(a): usize[a]
```
I’m ok with leaving the implementation like this for now as doing this reflection requires further design.

File: flux-desugar/src/desugar.rs (hunk 1 above)
Comment (ranjitjhala) [reply flux-desugar/src/desugar.rs:1089]: done!

File: flux-desugar/src/desugar.rs (hunk 4 above)
Comment (ranjitjhala) [reply flux-desugar/src/desugar.rs:887]: done!

File: flux-typeck/src/checker.rs (hunk 6 above)
Comment (ranjitjhala) [reply flux-typeck/src/checker.rs:710]: Yes, that was my recollection too -- let me remove!

**Comment by ranjitjhala** [comment]
Merging this in, let me make a separate issue for the `Array` thing.

**Pull request merged**